edi/models/edi_exchange_consumer_mixin.py

```diff
@@ -19,13 +25,114 @@ class EDIExchangeConsumerMixin(models.AbstractModel):
+        result = {}
+        for exchange_type in exchange_types:
+            eval_ctx = dict(
+                self._get_eval_context(), record=self, exchange_type=exchange_type
+            )
+            domain = safe_eval(exchange_type.enable_domain or "[]", eval_ctx)
+            if not self.filtered_domain(domain):
+                continue
+            if exchange_type.enable_snippet:
+                safe_eval(
+                    exchange_type.enable_snippet, eval_ctx, mode="exec", nocopy=True
+                )
+                if not eval_ctx.get("result", False):
+                    continue
+            result[exchange_type.id] = exchange_type.display_name
+        return result
+
+    def _get_eval_context(self):
+        """ Prepare the context used when evaluating python code
+            :returns: dict -- evaluation context given to safe_eval
+        """
+        return {
+            "datetime": datetime,
+            "dateutil": dateutil,
+            "time": time,
+            "uid": self.env.uid,
+            "user": self.env.user,
+        }
+
+    @api.model
+    def fields_view_get(
+        self, view_id=None, view_type="form", toolbar=False, submenu=False
+    ):
+        res = super().fields_view_get(
+            view_id=view_id, view_type=view_type, toolbar=toolbar, submenu=submenu
+        )
+        if view_type == "form":
+            doc = etree.XML(res["arch"])
+            for node in doc.xpath("//sheet"):
+                group = False
+                if hasattr(self, "_edi_generate_group"):
+                    group = self._edi_generate_group
+                str_element = self.env["ir.qweb"].render(
+                    "edi.edi_exchange_consumer_mixin_buttons", {"group": group},
+                )
+                node.addprevious(etree.fromstring(str_element))
+            View = self.env["ir.ui.view"]
+
+            # Override context for postprocessing
+            if view_id and res.get("base_model", self._name) != self._name:
+                View = View.with_context(base_model_name=res["base_model"])
+            new_arch, new_fields = View.postprocess_and_fields(self._name, doc, view_id)
+            res["arch"] = new_arch
+            res["fields"] = new_fields
+        return res
+
+    def _edi_create_exchange_record_vals(self, exchange_type):
+        return {
+            "model": self._name,
+            "res_id": self.id,
+        }
+
+    def _edi_create_exchange_record(self, exchange_type, backend):
+        exchange_record = backend.create_record(
+            exchange_type.code, self._edi_create_exchange_record_vals(exchange_type)
+        )
+        self._event("on_generate_manual_edi").notify(self, exchange_record)
+        return exchange_record.get_formview_action()
+
+    def edi_create_exchange_record(self, exchange_type_id):
+        self.ensure_one()
+        exchange_type = self.env["edi.exchange.type"].browse(exchange_type_id)
+        if exchange_type.backend_id:
```

Comment:
you can do the same if you find only one backend for exchange_type.backend_type_id :wink:

Reply:
OK